src/components/pushpin-code/Changes.tsx

```diff
@@ -0,0 +1,172 @@
+  getChanges,
+  applyChanges,
+  clone,
+  Patch,
+} from "@automerge/automerge"
+
+export function useAdvanceLastSeenHeads(docUrl: PushpinUrl) {
+  const workspaceId = useContext(WorkspaceContext)
+  const [, changeWorkspaceDoc] = useDocument<WorkspaceDoc>(workspaceId)
+  const [doc] = useDocument(parseDocumentLink(docUrl).documentId)
+
+  return useCallback(() => {
+    changeWorkspaceDoc((workspaceDoc) => {
+      if (!doc) {
+        return
+      }
+
+      let lastSeenHeadsByDocUrl = workspaceDoc.persistedLastSeenHeads
+      if (!lastSeenHeadsByDocUrl) {
+        workspaceDoc.persistedLastSeenHeads = {}
+        // need to read persistedLastSeenHeads from workspaceDoc again, so we get the proxied automerge object
+        lastSeenHeadsByDocUrl = workspaceDoc.persistedLastSeenHeads
+      }
+
+      lastSeenHeadsByDocUrl[docUrl] = getHeads(doc)
+    })
+  }, [doc, changeWorkspaceDoc])
+}
+
+const CURRENTLY_VIEWED_DOC_URLS: { [url: PushpinUrl]: boolean } = {}
+
+function isDocUrlCurrentlyViewed(url: PushpinUrl): boolean {
+  return CURRENTLY_VIEWED_DOC_URLS[url] ?? false
+}
+
+export function useAutoAdvanceLastSeenHeads(docUrl: PushpinUrl) {
+  const [doc] = useDocument(parseDocumentLink(docUrl).documentId)
+  const advanceLastSeenHeads = useAdvanceLastSeenHeads(docUrl)
+
+  useEffect(() => {
+    CURRENTLY_VIEWED_DOC_URLS[docUrl] = true
+
+    return () => {
+      delete CURRENTLY_VIEWED_DOC_URLS[docUrl]
+    }
+  }, [docUrl])
+
+  useEffect(() => {
+    advanceLastSeenHeads()
+  }, [doc])
+}
+
+export type LastSeenHeads = Heads | "latestHeads"
+
+export interface LastSeenHeadsMap {
+  [url: PushpinUrl]: LastSeenHeads
+}
+
+export interface PersistedLastSeenHeadsMap {
+  [url: PushpinUrl]: Heads
+}
+
+export function useLastSeenHeads(
+  docUrl: PushpinUrl
+): LastSeenHeads | undefined {
+  const workspaceId = useContext(WorkspaceContext)
+  const [workspaceDoc, changeWorkspaceDoc] =
+    useDocument<WorkspaceDoc>(workspaceId)
+
+  if (CURRENTLY_VIEWED_DOC_URLS[docUrl]) {
+    return "latestHeads"
+  }
+
+  if (!workspaceDoc || !workspaceDoc.persistedLastSeenHeads) {
+    return undefined
+  }
+
+  const lastSeenHeads = workspaceDoc.persistedLastSeenHeads[docUrl]
+
+  if (!lastSeenHeads) {
+    changeWorkspaceDoc((workspaceDoc) => {
+      let lastSeenHeadsByDocUrl = workspaceDoc.persistedLastSeenHeads
+      if (!lastSeenHeadsByDocUrl) {
+        workspaceDoc.persistedLastSeenHeads = {}
+        // need to read persistedLastSeenHeads from workspaceDoc again, so we get the proxied automerge object
+        lastSeenHeadsByDocUrl = workspaceDoc.persistedLastSeenHeads
+      }
+
+      lastSeenHeadsByDocUrl[docUrl] = []
+    })
+  }
+
+  return lastSeenHeads
+}
+
+export function getUnseenPatches(
+  doc: Doc<any>,
+  lastSeenHeads?: LastSeenHeads
+): Patch[] {
+  const patches: Patch[] = []
+
+  if (!hasDocUnseenChanges(doc, lastSeenHeads)) {
+    return []
+  }
+
+  try {
+    const oldDoc = clone(view(doc, lastSeenHeads as Heads))
```

Comment:
I'm sometimes getting an automerge exception here when calling clone. Am I using the API here incorrectly?
<img width="847" alt="Screen Shot 2022-12-01 at 11 38 52" src="https://user-images.githubusercontent.com/650540/205032030-c1b02004-1b91-496a-b98d-1dcb361e4560.png">

What's surprising to me is that the exception is thrown on the clone and not the view call. As far as I can tell it doesn't affect the behavior of the changed dots, they still get rendered correctly.

Reply:
@pvh

Reply:
I suspect that the "invalid hash" error message is slightly misleadingly named there. I think the problem is that `clone` throws an error if one of the hashes you pass to it is not in the document to be cloned, so it's likely that `lastSeenHeads` somehow contains a hash which is not in `oldDoc`.

Reply:
Does the view function not check the hashes and just accept invalid ones?

Reply:
It does not, it errors out. This is quite an unfriendly API which I want to improve. I don't think it should just silently accept invalid ones though because I think attempting to checkout heads which do not exist is probably a programming error a fair emount of the time.

Reply:
A better error message would be a good start for sure!